pkg/volume/iscsi/iscsi.go

```diff
@@ -93,109 +96,71 @@ func (plugin *iscsiPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
 	if err != nil {
 		return nil, err
 	}
-	tp, portals, iqn, lunStr, err := getISCSITargetInfo(spec)
+	iscsiDisk, err := createISCSIDisk(spec, podUID, plugin, manager, secret)
 	if err != nil {
 		return nil, err
 	}
+	return &iscsiDiskMounter{
+		iscsiDisk:    iscsiDisk,
+		fsType:       fsType,
+		readOnly:     readOnly,
+		mounter:      &mount.SafeFormatAndMount{Interface: mounter, Exec: exec},
+		exec:         exec,
+		deviceUtil:   ioutil.NewDeviceHandler(ioutil.NewIOHandler()),
+		mountOptions: volume.MountOptionFromSpec(spec),
+	}, nil
+}
 
-	lun := strconv.Itoa(int(lunStr))
-	portal := portalMounter(tp)
-	var bkportal []string
-	bkportal = append(bkportal, portal)
-	for _, p := range portals {
-		bkportal = append(bkportal, portalMounter(string(p)))
-	}
-
-	iface, initiatorNamePtr, err := getISCSIInitiatorInfo(spec)
-	if err != nil {
-		return nil, err
+func (plugin *iscsiPlugin) NewBlockVolumeMapper(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.BlockVolumeMapper, error) {
+	// If this is called via GenerateUnmapDeviceFunc(), pod is nil.
+	// Pass empty string as dummy uid since uid isn't used in the case.
+	var uid types.UID
+	var secret map[string]string
+	var err error
+	if pod != nil {
+		uid = pod.UID
+		secret, err = createSecretMap(spec, plugin, pod.Namespace)
+		if err != nil {
+			return nil, err
+		}
 	}
+	return plugin.newBlockVolumeMapperInternal(spec, uid, &ISCSIUtil{}, plugin.host.GetMounter(plugin.GetPluginName()), plugin.host.GetExec(plugin.GetPluginName()), secret)
+}
 
-	var initiatorName string
-	if initiatorNamePtr != nil {
-		initiatorName = *initiatorNamePtr
-	}
-	chapDiscovery, err := getISCSIDiscoveryCHAPInfo(spec)
+func (plugin *iscsiPlugin) newBlockVolumeMapperInternal(spec *volume.Spec, podUID types.UID, manager diskManager, mounter mount.Interface, exec mount.Exec, secret map[string]string) (volume.BlockVolumeMapper, error) {
+	readOnly, _, err := getISCSIVolumeInfo(spec)
 	if err != nil {
 		return nil, err
 	}
-	chapSession, err := getISCSISessionCHAPInfo(spec)
+	iscsiDisk, err := createISCSIDisk(spec, podUID, plugin, manager, secret)
 	if err != nil {
 		return nil, err
 	}
-
-	return &iscsiDiskMounter{
-		iscsiDisk: &iscsiDisk{
-			podUID:         podUID,
-			VolName:        spec.Name(),
-			Portals:        bkportal,
-			Iqn:            iqn,
-			lun:            lun,
-			Iface:          iface,
-			chap_discovery: chapDiscovery,
-			chap_session:   chapSession,
-			secret:         secret,
-			InitiatorName:  initiatorName,
-			manager:        manager,
-			plugin:         plugin},
-		fsType:       fsType,
-		readOnly:     readOnly,
-		mounter:      &mount.SafeFormatAndMount{Interface: mounter, Exec: exec},
-		exec:         exec,
-		deviceUtil:   ioutil.NewDeviceHandler(ioutil.NewIOHandler()),
-		mountOptions: volume.MountOptionFromSpec(spec),
+	return &iscsiDiskMapper{
+		iscsiDisk:  iscsiDisk,
+		readOnly:   readOnly,
+		mounter:    &mount.SafeFormatAndMount{Interface: mounter, Exec: exec},
```

Comment:
Since it is a raw block device, I am curious why mounter is instantiated with SafeFormatAndMount, should it not be just nil as it is completely up to the application running in the pod to decide what to do with the raw storage.

Reply:
@sbezverk We should remove mounter since it isn't used. thanks.